internal/events/aggregator_batch_state.go

```diff
@@ -0,0 +1,387 @@
+// Copyright © 2022 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"context"
+	"crypto/sha256"
+	"database/sql/driver"
+	"encoding/binary"
+
+	"github.com/hyperledger/firefly/internal/definitions"
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/pkg/database"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+func newBatchState(ag *aggregator) *batchState {
+	return &batchState{
+		database:           ag.database,
+		definitions:        ag.definitions,
+		maskedContexts:     make(map[fftypes.Bytes32]*nextPinGroupState),
+		unmaskedContexts:   make(map[fftypes.Bytes32]*contextState),
+		dispatchedMessages: make([]*dispatchedMessage, 0),
+
+		PreFinalize: make([]func(ctx context.Context) error, 0),
+		Finalize:    make([]func(ctx context.Context) error, 0),
+	}
+}
+
+// nextPinGroupState manages the state during the batch for an individual masked context.
+// We read it from the database the first time a context is touched within a batch, and then
+// replace the pins in it as they are spent/replaced for each pin processed.
+// At the end we flush the changes to the database
+type nextPinGroupState struct {
+	groupID           *fftypes.Bytes32
+	topic             string
+	nextPins          []*fftypes.NextPin
+	new               bool
+	identitiesChanged map[string]bool
+}
+
+type NextPinState struct {
+	nextPinGroup *nextPinGroupState
+	nextPin      *fftypes.NextPin
+}
+
+// contextState tracks the unmasked (broadcast) pins related to a particular context
+// in the batch, as they might block further messages in the batch
+type contextState struct {
+	blockedBy int64
+}
+
+// dispatchedMessage is a record for a message that was dispatched as part of processing this batch,
+// so that all pins associated to the message can be marked dispatched at the end of the batch.
+type dispatchedMessage struct {
+	batchID       *fftypes.UUID
+	msgID         *fftypes.UUID
+	firstPinIndex int64
+	lastPinIndex  int64
+}
+
+// batchState are synchronous actions to be performed while processing system messages, but which must happen after reading the whole batch
+type batchState struct {
```

Comment:
The description here doesn't fully cover the expanded purpose of this class

Reply:
👍  - I've added a bunch of extra text here - might be good for you to validate my understanding @awrichar

Reply:
Looks very good, thanks 🤩